### ISSUE TITLE
Rating Plugin does not appear in Constant Editor

### ISSUE THREAD

**Issue by jowe1** [thucke/TYPO3.ext.th_rating#6]
As in the documentation described I tried to add the PID for [plugin.tx_thrating.settings.pluginStoragePid] but whenever I open the constant editor the plugin does not appear under categories. I already included it in my Default Root Template and of course it is installed (via composer) and activated in Extension Manager.

I am using the latest version of Typo3 8.7.9 and I am running the site locally in my docker container currently. After everything is installed I wanted to deploy it with this rating plugin but if I can not set the storage page I wonder if it will be useless or not.

Any help is appreciated. Btw. I just updated from 7.6 LTS and everything else works fine, but maybe this bug does happen when updating. Database analyzer did not show anything, though (regarding ratings).

Thank you!

**Update: The plugin was not installed correctly (missing from LocalConfiguration and PackageStates.php). I wonder why that is because all other extensions were activated fine when adding to composer and composer update in terminal. Maybe the composer.json is not correctly built?**

Update2: Now that I included everything the plugin does not show anything in frontend except this error:

> Warning!
> Issue found with configuration for rating display - using default image. (1403192702)

I do not understand how to make it work and I already read the documentation and configured steps and almost clicked every option there is.

**Comment by thucke** [comment]
Thank you Joachim,
I'll have to setup a composer installation to check your issue

**Comment by thucke** [comment]
Hi @jowe1,
I cannot reproduce the mentioned issue.
Would you please provide me with your composer.json file you used to setup your site?

**Comment by jowe1** [comment]
Hello @thucke,
thanks for your quick reply.
`{
    "repositories": {
        "0": {
            "type": "composer",
            "url": "https://composer.typo3.org/"
        }
    },
    "name": "renoi/t3v8",
    "require": {
        "typo3/class-alias-loader": "^1.0.0",
        "typo3/cms": "^8.7",
        "typo3-ter/static-info-tables": "^6.4.3",
        "fluidtypo3/vhs": "^4",
        "dmitryd/typo3-realurl": "^2",
        "thucke/th-rating": "^1",
        "in2code/powermail": "^4",
        "matthiasmullie/minify": "^1",
        "instituteweb/min": "^1",
        "b13/seo_basics": "^0"
    },
    "autoload": {
        "classmap": [
            "typo3conf/ext/powermail/",
            "typo3conf/ext/neustacontentelements/",
            "typo3conf/ext/lightningsoulteaser/",
            "typo3conf/ext/disqus_comments_extbase/",
            "typo3conf/ext/realurl/",
            "typo3conf/ext/youtube_playlist/",
            "typo3conf/ext/tp3ratings/",
            "typo3conf/ext/subpageteaser/"
        ]
    },
    "extra": {
        "typo3/cms": {
            "cms-package-dir": "{$vendor-dir}/typo3/cms"
        }
    }
}
`
This is my composer.json. I also noticed that there was a mistake in the file tt_content.php in line 13:
`	\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addPiFlexFormValue($pluginSignature, 'FILE:EXT:th_rating/Configuration/Flexforms/flexform_pi1.xml');
`
It should not be th_rating with an underscore but a dash instead, otherwise in the backend it throws an error "invalid data structure, file not found".

As I fixed this manually I could add the plugin to a page but there is still no setup in the constants editor:
http://joachim-wetzel.de/img/upload/2017-12-30-12-34-02_th-rating-missing-constants-editor-small.jpg

Did I miss anything? I followed all the steps from your documentation and am using Typo3 v8.7.9.

Thank you!

**Comment by jowe1** [comment]
Now that I renamed the folder th-rating to th_rating it appears in the constants editor. I still get nothing in the frontend, though.
The only notification is:

> Notification
> FE user cannot be identified! (1403201096)

**Comment by thucke** [comment]
I will check if ther could be an issue with the packagist configuration.
In th measwhile would you please try 
`"typo3-ter/th-rating": "^1.0"`
in you composer file instead?

You're right - the partials are all with a non-capital letter.
I will open a new issue to solve that in the next version.

**Comment by thucke** [comment]
Regarding this one here:
> Notification
> FE user cannot be identified! (1403201096)
Would you please check if `plugin.tx_felogin_pi1.storagePid` points to the storage page where your website users reside?
Is it correct that you don't have any existing vote yet?
Did you activate anonymous voting?
It'd be great if you coud send me your current configuration array of th_rating.

**Comment by jowe1** [comment]
I do not really have any FE users so I created one and set the storagePid to a newly created folder. Now that I've installed typo3-ter/th-rating via composer there is a frontend rendering and I can also click the stars. But after clicking the network panel of Chrome shows a 404 XHR with the following message inside its response:

> 						<div class="media-body">
							<h4 class="alert-title">Page Not Found</h4>
							<div class="callout-body">
								Reason: Request parameters could not be validated (&amp;cHash comparison failed)
								
							</div>

Here is my configuration:
`plugin.tx_thrating.ratings {
	#use this section to configure your own ratingConfigurations
	exampleRatingConfig {
		imagefile = typo3conf/ext/th_rating/Resources/Public/Css/rating_bar.png
		barimage = 1
		tilt = 0
		settings {
			showNoFEUser = 1
			showNotRated = 1
			mapAnonymous = 1
			cookieLifetime = 0
		}
		fluid {
			layouts {
				default {
					hideSummary = {$plugin.tx_thrating.config.hideSummary}
					showCurrentRates = 1
					showSectionContent = 1
				}
			}
			templates {
				#ratinglinks.likesMode = 0 - deprecated from Version 0.10.2
			}
			partials {
				usersRating {
					showTextInfo = 1
					showGraphicInfo = 0
				}
				infoBlock {
					showAnonymousVotes = 1
				}
			}
		}
	}
}`

What I don't really understand is the backend field for "Enter value to activate cookie protection for anonymous votes". No matter what I enter it stays 0 and in the flexform it says 						<eval>trim,int,num,nospace</eval>
but I have no idea what kind of value it expects and for what as it is also set to size = 4 and max = 4?! Thanks again for your help!

**Comment by jowe1** [comment]
Ok, I just found the fix for this in the install tool, set this to false in All Configuration:
[FE][pageNotFoundOnCHashError] = false

Now, it is possible to rate on the site where I included the plugin (after fixing the Partial file names, of course). :) :+1:

**Comment by thucke** [comment]
Hmm,
regardless that rating is possible now it's still a strange behaviour.
I've never touched [pageNotFoundOnCHashError] and on my sites the AJAX requests were executed properly.
The value for cookie protection is nothing else that the cookieLifetime in days.
Is you website already online and reachable? Im interested in taking a glimpse view on the generated source code.
Which other infrastructure do you use (Operating system, Webserver)?

**Comment by thucke** [comment]
Hi @jowe1,
I've updated the extension that composer based installations now useing the correct directory name.
Is there still anything I can help to resolve?

**Comment by jowe1** [comment]
Hello @thucke,
thank you very much. No, I think my issue is solved by that.

But I would like to enhance the output a little bit (and have already done that) regarding google schema. Wouldn't it make sense to use a full schema in the output?

Here is the code:
```
<div class='tx-thrating-summary' itemscope itemtype="//schema.org/Article">
			<f:comment><!-- display check regarding rating summary information --></f:comment>
			<f:if condition="{settings.fluid.layouts.default.showCurrentRates}"><f:then>
				<meta itemprop="mainEntityOfPage" itemscope itemtype="http://schema.org/WebPage"/>
				<meta itemprop="author" content="Lightningsoul"/>
				<meta itemprop="datePublished" content="<f:cObject typoscriptObjectPath='lib.crdate' />"/>
				<meta itemprop="dateModified" content="<f:cObject typoscriptObjectPath='lib.lcdate' />"/>
				<meta itemprop="image" content="https://<f:cObject typoscriptObjectPath='lib.currentdomain' />/<f:cObject typoscriptObjectPath='lib.headerimage' />"/>
				<div itemprop="publisher" itemscope itemtype="https://schema.org/Organization">
					<meta itemprop="name" content="Renoi"/>
					<div itemprop="logo" itemscope itemtype="https://schema.org/ImageObject">
						<meta itemprop="url" content="https://renoi.de/typo3conf/ext/renoide/Resources/Public/img/renoi_de_banner_small_trans.png">
						<meta itemprop="width" content="500">
						<meta itemprop="height" content="65">
					</div>
				</div>
				<meta itemprop="description" content="<f:cObject typoscriptObjectPath='lib.pageDescription' />"/>
				<meta itemprop="headline" content="<f:cObject typoscriptObjectPath='lib.pageTitle' />"/>
		
		<span data-role="onCurrentRateActive" style="{f:if(condition: rating.currentRates.currentrate, then: '', else: 'display:none;')}" itemprop="aggregateRating" itemscope itemtype="//schema.org/AggregateRating">
					<meta itemprop="name" content="<f:translate key='fluid.layout.default.description'/>"/>
					<f:format.raw><f:translate key='fluid.layout.default.currentRating' default='fluid.layout.default.currentRating' arguments="{1:rating.currentRates.currentrate, 2:'{f:count(subject:\"{rating.votes}\")}'}"/></f:format.raw>
					<f:if condition="{anonymousVotes}"><f:then>
					<br/><f:format.raw><f:translate key='fluid.layout.default.anonymousVotes' default='fluid.layout.default.anonymousVotes' arguments='{1:anonymousVotes}'/></f:format.raw>
					</f:then></f:if>
					<meta itemprop="bestRating" content="{stepCount}"/>
					<meta itemprop="worstRating" content="1"/>
				</span>
```
DefaultTemplate.html lines 14-40. The new libraries need to be filled in Typoscript, of course (I can send you the code as well, in case you are interrested).

Thanks again.

**Comment by thucke** [comment]
Hi @jowe1,
I'd appreciate if you could send me your suggested code changes.
Thank you in advance!

**Comment by jowe1** [comment]
Add this to th_rating/Configuration/TypoScript/setup.txt:
```
lib.pageTitle = TEXT
lib.pageTitle.data = page:title

lib.pageDescription = TEXT
lib.pageDescription.data = page:description

lib.author = TEXT
lib.author.value = page:author

lib.currentURL=TEXT
lib.currentURL.typolink {
    parameter.data=TSFE:id
    returnLast=url
}

lib.currentdomain = TEXT
lib.currentdomain {
    data = getEnv:HTTP_HOST
}

lib.crdate = TEXT
lib.crdate {
    data = page : crdate
    date = c
}

lib.lcdate = TEXT
lib.lcdate {
    data = page : SYS_LASTCHANGED
    date = c
}

lib.headerimage = IMG_RESOURCE
lib.headerimage {
    file {
        import.data = levelmedia:-1, slide
        treatIdAsReference = 1
        import.listNum = 0
    }
}
```
And apply the changes to DefaultTemplate.html mentioned above. In addition "Renoi" should be set to a constant or database entry (maybe from constant editor).

**Comment by thucke** [comment]
Hi @jowe1,

thank you a lot for your suggestions.
I try to understand you modifications before I include them in the distibution.
Actually I am moved by the thought if all modifications hitting the sense of the rating entries. Let me try to explain - I appreciate you comments:

The extension doesn't generate only one page. It generated multiple page objects so called "ratings". On some websites (e.g. product listings) it is possible that really a lot of rating instances will be included in the output.
That's why I'm not sure if general meta information like `mainEntityOfPage`, `author`, `datePublished` and so on should be included in EVERY instance. Additionally you changed to `itemtype="http://schema.org/Article"` instead of `Thing`. Would you see it legal against the aim of rich snippets to include this information multiple times on one page? It'll be redundant, wouldn't it?
Next point would be that the transferred data volume would also raise without any need.

It think it'd be worse to take a deeper look how the extension implements rich snippets.
Perhaps I will have gain more knowlegde on that specific technology.

Basically the combination with default typoscript libs seems very attractive to me.
